catalog/app/utils/useQuery.ts

```diff
@@ -0,0 +1,39 @@
+import * as R from 'ramda'
+import * as urql from 'urql'
+
+import useMemoEq from 'utils/useMemoEq'
+
+interface OpCases<D, DR, ER, FR> {
+  data: (data: D, error: urql.CombinedError | undefined) => DR
+  error: (error: urql.CombinedError | undefined) => ER
+  fetching: () => FR
+}
+
+const opCase =
+  <D>(result: urql.UseQueryState<D, any>) =>
+  <DR, ER, FR>({ data, error, fetching }: OpCases<D, DR, ER, FR>) => {
+    if (result.fetching) return fetching()
+    if (result.data) return data(result.data, result.error)
+    return error(result.error)
+  }
+
+interface UseQueryArgs<V, D> extends urql.UseQueryArgs<V, D> {
+  suspend?: boolean
+}
+
+export interface UseQueryResult<D, V = any> extends urql.UseQueryState<D, V> {
+  case: <DR, ER, FR>(cases: OpCases<D, DR, ER, FR>) => DR | ER | FR
+  run: (opts?: Partial<urql.OperationContext>) => void
+}
+
+export function useQuery<V, D>({
```

Comment:
In typescript docs, they use one-letter types because they use abstract examples. But these are regular variables, and need the same naming as usual: `V → Value, D → ?`